aws/resource_aws_sagemaker_endpoint.go

```diff
@@ -0,0 +1,224 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/sagemaker"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsSagemakerEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSagemakerEndpointCreate,
+		Read:   resourceAwsSagemakerEndpointRead,
+		Update: resourceAwsSagemakerEndpointUpdate,
+		Delete: resourceAwsSagemakerEndpointDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				ValidateFunc: validateSagemakerName,
+			},
+
+			"endpoint_config_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateSagemakerName,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsSagemakerEndpointCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else {
+		name = resource.UniqueId()
+	}
+
+	createOpts := &sagemaker.CreateEndpointInput{
+		EndpointName:       aws.String(name),
+		EndpointConfigName: aws.String(d.Get("endpoint_config_name").(string)),
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		createOpts.Tags = tagsFromMapSagemaker(v.(map[string]interface{}))
+	}
+
+	log.Printf("[DEBUG] SageMaker endpoint create config: %#v", *createOpts)
+	_, err := conn.CreateEndpoint(createOpts)
+	if err != nil {
+		return fmt.Errorf("error creating SageMaker endpoint: %s", err)
+	}
+
+	d.SetId(name)
+	log.Printf("[INFO] SageMaker endpoint ID: %s", d.Id())
+
+	log.Printf("[DEBUG] Waiting for SageMaker endpoint (%s) to become available", d.Id())
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"Creating"},
+		Target:  []string{"InService"},
+		Refresh: SagemakerEndpointStateRefreshFunc(conn, d.Id()),
+		Timeout: 15 * time.Minute,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf(
+			"error while waiting for SageMaker endpoint (%s) to become available: %s",
+			d.Id(), err)
+	}
+
+	return resourceAwsSagemakerEndpointRead(d, meta)
+}
+
+func resourceAwsSagemakerEndpointRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	endpointRaw, _, err := SagemakerEndpointStateRefreshFunc(conn, d.Id())()
```

Comment:
We should prefer to use `conn.DescribeEndpoint()` here so we can allow all endpoint statuses, e.g.

```
const (
    // EndpointStatusOutOfService is a EndpointStatus enum value
    EndpointStatusOutOfService = "OutOfService"

    // EndpointStatusCreating is a EndpointStatus enum value
    EndpointStatusCreating = "Creating"

    // EndpointStatusUpdating is a EndpointStatus enum value
    EndpointStatusUpdating = "Updating"

    // EndpointStatusSystemUpdating is a EndpointStatus enum value
    EndpointStatusSystemUpdating = "SystemUpdating"

    // EndpointStatusRollingBack is a EndpointStatus enum value
    EndpointStatusRollingBack = "RollingBack"

    // EndpointStatusInService is a EndpointStatus enum value
    EndpointStatusInService = "InService"

    // EndpointStatusDeleting is a EndpointStatus enum value
    EndpointStatusDeleting = "Deleting"

    // EndpointStatusFailed is a EndpointStatus enum value
    EndpointStatusFailed = "Failed"
)
```

If necessary, we can remove the resource from the state afterwards with a check like:

```go
if aws.StringValue(endpoint.EndpointStatus) == sagemaker.EndpointStatusDeleting {
  log.Printf("[WARN] SageMaker Endpoint (%s) is deleting, removing from state", d.Id())
  d.SetId("")
  return nil
}
```